helix-core/src/selection.rs

```diff
@@ -1185,25 +1183,22 @@ mod test {
                 // TODO: rather than this behavior, maybe we want it
                 // to be based on which side is the anchor?
                 //
-                // We get a leading zero-width range when there's
-                // a leading match because ranges are inclusive on
-                // the left.  Imagine, for example, if the entire
-                // selection range were matched: you'd still want
-                // at least one range to remain after the split.
-                Range::new(0, 0),
+                // Ranges returned are exclusive on the left and inclusive on the right. Logic is
+                // as follows: Continue down the string until a match is found or the end is
+                // reached, this point will be the anchor for the current range. The head of
+                // the previous range is the start of the current match. Any anchors should
+                // come directly after a match
                 Range::new(1, 5),
                 Range::new(6, 9),
                 Range::new(11, 13),
                 Range::new(16, 19),
-                // In contrast to the comment above, there is no
-                // _trailing_ zero-width range despite the trailing
-                // match, because ranges are exclusive on the right.
+                Range::new(20, 20),
```

Comment:
I believe we don't want this trailing point either. With (essentially) the same reproduction case as #3850: `i| | | |<esc>%_S\|<ret>`: we have a trailing selection created at the end which lives outside of the selection at the start of `S` (when the selection is over `| | | |`). I believe that case should leave only the spaces between the `|`s selected.